### ISSUE TITLE
Add Option to Allow Re-Using Opened SqlConnection

### ISSUE THREAD

**Issue by AlexVPerl** [roadrunnerlenny/etlbox#39]
First of, great work on this library!

ControlFlow allows to set 1 current DbConnection, so if one needs to have multiple threads doing work on separate DBs / Connections, have to pass SqlConnectionManager instance to each method directly, example: 

`DropDatabaseTask.DropIfExists(new SqlConnectionManager("connection string"), "testDb");`

Internally I see that ETLBox will clone an instance of SqlConnectionManager and then open a new connection every time. This behavior may be desirable. However this is also a performance hit. 

It would be also desirable to have an option to pass already opened SqlConnection instance to avoid cloning and opening new connection every time. This way an SqlConnection can be re-used for performance benefit.

**Comment by roadrunnerlenny** [comment]
Hi Alex,

thanks for your feedback and your interest in ETLBox! I appreciate it, and hope that the library does actually help you.  

Regarding your question, I am not quite sure if I really understand what your intention is. 

First of all, please let me clarify something about the Open() method of the connection manager: Whenever this method is called (doesn't matter if this happens on a completely new object or the object already exists), it will use instantiate a new connection on the underlying ADO.NET provider. It will also call the open method on this provider. *But this won't always create a new database connection*. The method naming `Open` could be misleading here - it actually should be something like 'OpenOrReuseConnectionFromPool`. The reason for that is that ADO.NET (which the whole connection handling is based upon) has it's own connection pooling. Whenever you try to open a connection (which includes opening up the right port, doing a handshake, authentification etc.), the ADO.NET connection pooling will check if it makes sense to open up a new connection or if it could reuse an existing - already opened - connection from the pool. Whenever you call the `Close()` method (which is totally recommended when using ADO.NET),this won't automatically close the connection. It will just return the connection to the pool - it will stay open until ADO.NET decides that it not needed anymore. 

Please see here for more details of connection pooling: https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-connection-pooling

Thus having said, I don't think that calling the open / close (and also recreating the object every time) is actually a performance hit. It will just use the connection pooling provided - as recommended by ADO.NET, which probably will perform already very well. So if you would create a test environment where you open up a lot of connections to the database, you would see that the connection pool would have only one connection which stays open. So calling `Open` and `Close` multiple times in a row would in most cases never open up more than one connection. I also have a Test regarding the this pooling behavior in my test project:
https://github.com/roadrunnerlenny/etlbox/blob/91350599bbce378dc328f61749843bcedd3bf822/TestsETLBox/src/DataFlowTests/ConnectionManager/SqlConnectionManagerTests.cs#L64-L76

In this test class you will also find other scenarios testing the ADO.NET pooling.  

TL;DR; - the current implementation follows the recommendation of Microsoft and makes use of the connection pooling provided by ADO.NET. This means that this actually can increase your performance, and in most scenarios you never have more connections open that you actually need for your application. 

Sorry for writing so much, but now these are the question I actually have about your feature request: 
- In which scenario would you expect a performance hit? 
If you want to pass an already opened connection, you actually need to have the connection pooling disabled, otherwise you would just pass the exact same connection which would have been used anyways when trying to open the connection. Is this what you where intending?

Also, connection pooling relies that you always pass the same connection string. 
- Do you have a scenario in mind where you pass a lot of different connection string? 

And the last question I have in my mind is: 
- How would you actually like to pass the already opened connection object to the connection manager?

**Comment by roadrunnerlenny** [comment]
#1

**Comment by AlexVPerl** [comment]
Thank you Lenny for well thought out answer. 

1. You're right about connection pooling, ADO.NET does handle that part for us. I guess the only performance hit (possibly somewhat negligible) would come from cloning the object itself and ADO.NET's internal logic for re-using the connection. When you're running many (1000's or 10,000's) DB operations on the same connection just seems unnecessary for those steps to be repeated. 

2. For second question, I meant re-using the same connection with the same connection string.

3. Perhaps something like, IsOpened : bool flag in SqlConnectionManager class, where if set to true, cloning & opening statements would be skipped. 

Again, with ADO's Connection Pooling I agree with you, this is not an urgent issue. But also I think this change could be small enough to allow this flexibility to give user some control over the connection handling. Looking forward to your thoughts on this suggestion.

Thanks!

**Comment by AlexVPerl** [comment]
Quick follow up. So this left me wanting to run some benchmarks to gauge the difference between **Reusing Existing Connection vs. Using Connection Pooling** every time. Hoping this will be useful for curiosity's sake :)

I ran total of 8 different test runs. Each test run executes 5000 iterations of connection pooled code and 5000 iterations of reusing existing connection. 

For CPU I disabled Intel's SpeedStep, so CPU was constantly clocked at 100% - this way results aren't skewed by CPU's dynamic frequency scaling. Benchmark also sleeps the first 10 seconds so CPU settles after build and is at 0% usage when code starts running.

**Basically relying on ADO's connection pooling by opening a new connection every time has about 7.69% overhead vs using existing connection.** [Not terrible, not great](https://www.youtube.com/watch?v=wB7F_d_kwIg) :P

This also does not account for cloning the object itself. 

**Code used for benchmark:**

```

public static void TestSingleConnection(int pMaxIterations)
        {
            Stopwatch sw = new Stopwatch();
            sw.Start();
            System.Console.WriteLine("Starting TestSingleConnection() with {0} iterations", pMaxIterations);

            SqlConnection conn = new SqlConnection("<removed>");
            conn.Open();
            
            for (int i = 0; i < pMaxIterations; i++)
            {
                using var cmd = new SqlCommand("SELECT DISTINCT service_id FROM [GTFS_master].[dbo].[calendar_master]", conn);
                cmd.ExecuteNonQuery();
            }
            
            conn.Close();
            
            sw.Stop();
            System.Console.WriteLine("Finished TestSingleConnection() with {0} iterations. Time = {1}", pMaxIterations, sw.ElapsedMilliseconds); 
        }


        public static void TestConnectionPooling(int pMaxIterations)
        {
            Stopwatch sw = new Stopwatch();
            sw.Start();
            System.Console.WriteLine("Starting TestConnectionPooling() with {0} iterations", pMaxIterations);

            for (int i = 0; i < pMaxIterations; i++)
            {
                SqlConnection conn = new SqlConnection("<removed>");
                conn.Open();
                using var cmd = new SqlCommand("SELECT DISTINCT service_id FROM [GTFS_master].[dbo].[calendar_master]", conn);
                cmd.ExecuteNonQuery();
                conn.Close();
            }

            sw.Stop();
            System.Console.WriteLine("Finished TestConnectionPooling() with {0} iterations. Time = {1}", pMaxIterations, sw.ElapsedMilliseconds); 
        }

        public static void PrimeConnection()
        {
            Stopwatch sw = new Stopwatch();
            sw.Start();
            System.Console.WriteLine("PrimeConnection started...");

            using SqlConnection conn = new SqlConnection("<removed>");
            conn.Open();
            using var cmd = new SqlCommand("SELECT DISTINCT service_id FROM [GTFS_master].[dbo].[calendar_master]", conn);
            cmd.ExecuteNonQuery();
            conn.Close();

            sw.Stop();
            System.Console.WriteLine("PrimeConnection finished in {0}", sw.ElapsedMilliseconds);
        }


        static void Main(string[] args)
        {
            System.Console.WriteLine("Start ...");

            // Sleep for 10 seconds, allow CPU to settle and be at 0% load after build
            Thread.Sleep(10000); 

            // Prime connection, so at least 1 connection exists in the pool
            PrimeConnection();
            
            TestConnectionPooling(1000);
            TestConnectionPooling(1000);
            TestConnectionPooling(1000);
            TestConnectionPooling(1000);
            TestConnectionPooling(1000);
            
            TestSingleConnection(1000);
            TestSingleConnection(1000);
            TestSingleConnection(1000);
            TestSingleConnection(1000);
            TestSingleConnection(1000);
        }

```

**Results:**

![2019-11-12_0201](https://user-images.githubusercontent.com/6545404/68651459-75519180-04f5-11ea-8078-2ad60019bbc4.png)

**Comment by roadrunnerlenny** [comment]
Hey Alex,

thanks for looking into this! This is actually a quite interesting result that you come up with. So basically by always Open/Close the connection before every statement things are slowed down by about 8%. So returning the connection to the pool and then reusing it does take a little time - obviously. I would have expected that the pooling of ADO.NET worked faster, but you proved me wrong.

So you convinced me that there is actually a possibility of a performance leak when cloning the object and therefore closing and reopening the connection. 
I will investigate to add a property `LeaveOpen` to the ConnectionManager, so that it will keep the connection alive (something like to open it the first time and then never closing it). Probably there also needs to be a `Close` command exposed. 

By the way, there is actually a good reason why I am cloning the connection manager object and open the connection every time in the ConnectionManagers: This way it works when you have parallelization. Let's say you want to execute two `SqlTasks` in parallel, then you actually need to have two connections - or you need some kind of synchronization between both Tasks to figure out which is executed first. By cloning the object, I get every time a new fresh object, and the connection pooling from ADO.NET can deal with the connection handling. Basically it will try to create a new connection for every request running in parallel, or if the pool "is full" it will wait some time to get an existing one which was released back into the pool. 

Having this written, I will try to add some tests to simulate that behavior (parallel execution, and also perhaps a performance test for the new property `LeaveOpen` based on your code. )

I will come back to you when I was able to look into this and let you know about the progress. Of course,  if you already have an idea what to change where, feel free to post it here or to open up a PR!

**Comment by AlexVPerl** [comment]
That's great to hear! Looking forward to it :)

**Comment by roadrunnerlenny** [comment]
Hei Alex,

I just released version 1.6.4, which now contains a `LeaveOpen`property on every `DBConnectionManager` (basically all connection managers).
If set to `true` (default is `false`) , it will only open a new connection for the first time - use. 

So you could create a new ConnectionManager like this:
```
SqlConnectionManager con = new SqlConnectionManager("..connectionString...")
{
    LeaveOpen = true
};
```

If you use the connection manager the first, e.g. when you use a `SqlTask` like: 
```C#
SqlTask.ExecuteNonQuery(con, "Dummy", "SELECT 1");
```
it will then open up a new connection and leave it open. Whenever you call a task that uses the same connection, it will use the already opened connection.  

Whenever you are done with connection, don't forget to close it with:
```
con.Close();
```

Be aware that you can't use the same connection in parallel, or once the connection is broken or in a failed it can't be used any more.
I also added a property `ConnectionState` to every connection manager, so that you can query the state of a connection. 

Some examples are also here:
https://github.com/roadrunnerlenny/etlbox/blob/dev/TestsETLBox/src/DataFlowTests/ConnectionManager/SqlConnectionManagerLeaveOpenTests.cs

I hope this is what you expected?

**Comment by AlexVPerl** [comment]
This is awesome, thank you!

Also, excited about the MemorySource & MemoryDestination features - can't wait to try that out.
